util/tlgen/xbar.rtl.sv.tpl

```diff
@@ -154,18 +157,36 @@ module xbar_${xbar.name} (
     dev_sel_${block.name} = ${"%d'd%d" % (sel_len, len(block.ds))};
 % for i in block.ds:
 <%
+  # TODO: Add Xbar address handling
   leaf = xbar.get_leaf_from_s1n(block, loop.index);
   name_space = "ADDR_SPACE_" + leaf.name.upper();
   name_mask  = "ADDR_MASK_" + leaf.name.upper();
+  prefix = "if (" if loop.first else "end else if ("
 %>\
-  % if loop.first:
-    if ((${addr_sig} & ~(${name_mask})) == ${name_space}) begin
+  % if leaf.xbar == False:
+    ${prefix}(${addr_sig} & ~(${name_mask})) == ${name_space}) begin
+      dev_sel_${block.name} = ${"%d'd%d" % (sel_len, loop.index)};
+    ${"end" if loop.last else ""}
   % else:
-    end else if ((${addr_sig} & ~(${name_mask})) == ${name_space}) begin
-  % endif
+    ## Xbar device port
+<%
+  num_range = len(leaf.addr_range)
+%>\
+    ${prefix}
+    % for i in range(num_range):
+<%
+  # TODO: Check if the size is power of 2, if not convert into >= <=
+%>\
+      % if lib.is_pow2(leaf.addr_range[i][1]+1):
```

Comment:
this is great.  Is there any reason we shouldn't add this to devices also? I recall coming across places where supporting both power of 2 and non-powers of 2 was useful.

Reply:
Could you please explain more about the "add this to devices" ?

Reply:
i'm sorry, i meant adding the non-power-of-2 option for nodes that directly act as devices, and not only for xbar.  Or maybe I mis-read, and that is already the case (sorry unable to check right now).

Reply:
Updated